apollo-gradle-plugin-kotlin/src/main/kotlin/com/apollographql/apollo/gradle/kotlin/ApolloExtension.kt

```diff
@@ -0,0 +1,117 @@
+package com.apollographql.apollo.gradle.kotlin
+
+import com.apollographql.apollo.compiler.NullableValueType
+import org.gradle.api.Action
+
+open class ApolloExtension {
+
+  var nullableValueType = NullableValueType.ANNOTATED.value
+  var useSemanticNaming = true
+  var generateModelBuilder = false
+  var useJavaBeansSemanticNaming = false
+  var suppressRawTypesWarning = false
+  var generateKotlinModels = false
+  var generateVisitorForPolymorphicDatatypes = false
+  /**
+   * Deprecated, use services instead
+   */
+  var schemaFilePath: String? = null
+  /**
+   * Deprecated, use services instead
+   */
+  var outputPackageName: String? = null
+  var customTypeMapping: Map<String, String> = emptyMap()
+
+  var services = Services()
+  var sourceSet = SourceSet()
+
+  /**
+   * Deprecated
+   */
+  fun sourceSet(action: Action<SourceSet>) {
+    action.execute(sourceSet)
+  }
```

Comment:
why it's deprecated?

Reply:
I'm not super comfortable with the sourceSet block. Things that bug me:

* I find the "sourceSet" naming weird. Does it relate to .graphql files or schema.json ? 
* Since we allow multiple schema.json, we should allow multiple "sourceSet"
* The packageName should be tied to a schema.json. Or computed automatically from each source file relative directory ? 

This is why I'm trying to introduce the `Service` class, similar to a `retrofit` service. A service would have:

* a schema.json
* a list of .graphql files
* it's own include/exclude filters

Reply:
Originally `sourceSet ` was designed to give a support to provide explicitly schema location (only one schema for now, to support multiple we can support multiple  source sets) and exclude / include filters. It's related to both schema and GraphQL files that required (as source) to generate models.

In your case, could you please provide how is it going to look like in `gradle.build` plugin configuration section, to configure schema location, filters and target package?

Reply:
This is the part I'm not sure about. Do you know if anyone has graphql files in multiples subfolders ? Like so:

    src/main/graphql/com/example/schema.json
    src/main/graphql/com/example/movies/MovieQueries.graphql
    src/main/graphql/com/example/characters/CharactersQueries.graphql    

Being able to specify a single `outputPackageName` in that case feels weird as I would expect 

    com.example.movies.MovieQueries.kt
    com.example.characters.CharacterQueries.kt

Reply:
Currently, I'm thinking about exposing something like this: 

    apollo {
        services {
            create("githunt") {
                schemaFilePath = "graphql/com/githunt/schema.json"
                graphqlFilesFolder = "graphql/com/githunt/"
                exclude = listOf("README\\.md")
                // leave the default
                //include = listOf("*\\.graphql")
                // append apollo to not clash with other classes
                packageName = "com.githunt.apollo"
            }
            create("starwars") {
                schemaFilePath = "graphql/com/starwars/schema.json"
                // use defaults for everything else
            }
        }
    }

This obviously doesn't work if some users have multiple subfolders

Reply:
I updated the proposal above a bit. I think the most important change is that `packageName` should only affect the shared classes, not the individual classes. We can always map the current configuration and provide compatibility methods that would print a deprecation warning but I think this would be a good thing moving forward.

Reply:
> Do you know if anyone has graphql files in multiples subfolders ? 
```
src/main/graphql/com/example/schema.json
src/main/graphql/com/example/movies/MovieQueries.graphql
src/main/graphql/com/example/characters/CharactersQueries.graphql 
```

Yeah I saw couple usages  of that, for example to group queries by feature.

> Being able to specify a single `outputPackageName` in that case feels weird as I would expect

I believe there is a bug to fix this, as you are right the `outputPackageName` should be a as a root package. 
https://github.com/apollographql/apollo-android/issues/1367

Agree on your comment / proposal there to rename `packageName ` to `rootPackageName `.

> `services {`

Is there any way to flatten it a little bit, like:
```
apollo {
    service("githunt") {
            // Places to look for graphql files relative to the current sourceSet (e.g src/main/{sourceFolder})
            // You can place .graphql or .gql files in sub directories to use different namespaces
            // By default, the plugin looks in the "graphql" directory
            sourceFolders = listOf("graphql/githunt")
            // Place where the schema file is relative to the current project. For multi-module setups, this can go outside the module but in
            // that case, you'll want to define packageName too else everything will end up at the root of the namespace
            // By default, the plugin will look for a schema.json file in sourceFolders.
            schemaFile = "../common/graphql/src/main/graphql/githunt/com/githunt/schema.json"
            // The package name used for common classes (Custom Types and Fragments).
            // By default, this is computed from the path of schemaFile relative to sourceFolders or the empty string if it the schemaFile is not a child of sourceFolders.
            // Note that this does not affect the packageName of queries and mutation.
            packageName = "com.githunt.apollo"
            // list of files to exclude
            exclude = listOf("**/schema.*")
        }
        service("starwars") {
            sourceFolders = listOf("graphql/starwars")
            // use defaults for everything else
        }
}
```

Another question `service("githunt")` I assume that `githunt` is a name, where it's going to be used?

Reply:
> Agree on your comment / proposal there to rename packageName to rootPackageName 

If we want to make it 100% backward compatible, we'll have to keep both for some time. I'll investigate this.

> Is there any way to flatten it a little bit 

Yes, this is also what SQL Delight is doing. The reason for the "create" is that a lot of apis come from gradle where everything is a `NamedDomainObjectContainer` so in groovy you could just write `githunt()` instead of `create("githunt")` and it would add an object named "githunt" automagically.

We could do that as well (See [here](https://github.com/square/sqldelight/blob/bc5a8a3d3a6169a4d73efdc4b54fb215e3ccfd55/sqldelight-gradle-plugin/src/main/kotlin/com/squareup/sqldelight/gradle/SqlDelightExtension.kt#L27) for the magic):

    apollo {
        githunt {
            sourceFolders
            ...
        }
        starwars {
            ...
        }
        generateKotlinModels = true
    }

I'm personally not a fan as it mixes types but this might appeal to groovy users.

Reply:
About the "githunt" name, it's not actually used, it just serves as an identifier, like `signingConfigs` or `flavors` in android builds. Right now I'm using it in the generated output folder but that's about it...
I don't mind too much removing it altogether:

    apollo {
        service {
            sourceFolders = 
            ...
        }
        service {
            ...
        }
        generateKotlinModels = true
    }

But it feels a bit weird